service/storage_proxy.cc

```diff
@@ -1729,6 +1729,9 @@ class storage_proxy::cancellable_write_handlers_list : public bi::list<abstract_
         for (auto& itp : _live_iterators) {
             if (&**itp == handler) {
                 ++*itp;
+                if (*itp == end()) {
+                    unregister_live_iterator(itp);
```

Comment:
i don't think it's safe to remove an element in an `std::vector` while iterating over the vector using the range-based for loop, as the `iterator` pointing the erased element would be invalidated by `_live_iterators.erase()`. how about:

```c++
for (auto it = _live_iterators.begin(); it != _live_iterators.end();) {
    auto& itp = *it;
    if (auto& current_handler = **itp; &current_handler == handler) {
        ++*itp;
        if (*live_iterator == end()) {
            it = _live_iterators.erase(boost::remove(_live_iterators, itp), _live_iterators.end());
        } else {
            ++it;
        }
    } else {
        ++it
    }
}
```

Reply:
Right. But I think it is easier to add one more pass. `boost::remove` does it anyway and the array is usually has only one element anyway.

Reply:
agreed. i was uncertain about the vector's size, which complicates the approach proposed above. If we can accommodate a two-pass solution, it would surely significantly improve readability and maintainability of the code.